cirq/work/observable_measurement.py

```diff
@@ -191,21 +238,44 @@ def _check_meas_specs_still_todo(
     removing `meas_spec`s from the loop if they are done.
     """
     still_todo = []
-    repetitions: Set[int] = set()
+    repetitions_set: Set[int] = set()
     for meas_spec in meas_specs:
         accumulator = accumulators[meas_spec]
-        more_repetitions = _repetitions_to_do(accumulator, desired_reps=desired_repetitions)
+        more_repetitions = stopping_criteria.more_repetitions(accumulator)
+
+        if more_repetitions < 0:
+            raise ValueError(
+                "Stopping criteria's `more_repetitions` should return 0 or a positive number."
+            )
         if more_repetitions == 0:
             continue
 
-        repetitions.add(more_repetitions)
+        repetitions_set.add(more_repetitions)
         still_todo.append(meas_spec)
 
     if len(still_todo) == 0:
         return still_todo, 0
 
-    reps = _aggregate_n_repetitions(repetitions)
-    return still_todo, reps
+    repetitions = _aggregate_n_repetitions(repetitions_set)
+    total_repetitions = len(still_todo) * repetitions
+    if total_repetitions > MAX_REPETITIONS_PER_JOB:
+        old_repetitions = repetitions
+        repetitions = MAX_REPETITIONS_PER_JOB // len(still_todo)
+
+        if repetitions < 10:
+            raise ValueError(
+                "You have requested too many parameter settings to batch your job effectively. "
+                "Consider fewer sweeps or manually splitting sweeps into multiple jobs."
+            )
+
+        # coverage: ignore
+        warnings.warn(
```

Comment:
Recommend adding coverage for this warning with `pytest.warns`

Reply:
added, now that I know the monkeypatching trick for max_repetitoins_per_job :)